File: src/algorithms/greedy.js
// Step Generators for Greedy Algorithms

export function generateActivitySelectionSteps(startTimes = [1, 3, 0, 5, 8, 5], finishTimes = [2, 4, 6, 7, 9, 9]) {
  const n = startTimes.length
  const activities = Array.from({ length: n }, (_, i) => i).sort((a, b) => finishTimes[a] - finishTimes[b])
  const selected = []
  const steps = []

  steps.push({
    array: finishTimes,
    comparing: [],
    swapped: false,
    line: 1,
    description: 'Activity Selection: Sorting activities by finish time.'
  })

  let lastFinish = -1
  for (const idx of activities) {
    const st = startTimes[idx]
    const ft = finishTimes[idx]
    if (st >= lastFinish) {
      selected.push(idx)
      lastFinish = ft
      steps.push({
        array: [...finishTimes],
        comparing: [idx],
        swapped: true,
        sorted: [...selected],
        line: 4,
        description: `Selected Activity ${idx + 1} (Start: ${st}, Finish: ${ft})`
      })
    } else {
      steps.push({
        array: [...finishTimes],
        comparing: [idx],
        swapped: false,
        sorted: [...selected],
        line: 6,
        description: `Skipped Activity ${idx + 1} (Start: ${st} < Last Finish: ${lastFinish})`
      })
    }
  }

  steps.push({
    array: [...finishTimes],
    comparing: [],
    swapped: false,
    sorted: [...selected],
    line: 8,
    description: `Greedy Activity Selection Complete! Selected ${selected.length} non-overlapping activities.`
  })

  return steps
}

export function generateFractionalKnapsackSteps() {
  const weights = [10, 20, 30]
  const values = [60, 100, 120]
  const ratios = values.map((v, i) => v / weights[i])
  const capacity = 50
  let totalVal = 0
  let currCap = capacity
  const steps = []

  steps.push({
    array: [60, 100, 120],
    comparing: [],
    swapped: false,
    line: 1,
    description: 'Fractional Knapsack: Calculate Value-to-Weight ratios.'
  })

  const items = [0, 1, 2].sort((a, b) => ratios[b] - ratios[a])
  for (const idx of items) {
    const w = weights[idx]
    const v = values[idx]
    if (w <= currCap) {
      currCap -= w
      totalVal += v
      steps.push({
        array: [60, 100, 120],
        comparing: [idx],
        swapped: true,
        line: 4,
        description: `Took 100% of Item ${idx + 1} (Weight: ${w}, Value: ${v}). Remaining Capacity: ${currCap}`
      })
    } else {
      const fraction = currCap / w
      totalVal += v * fraction
      steps.push({
        array: [60, 100, 120],
        comparing: [idx],
        swapped: true,
        line: 6,
        description: `Took ${Math.floor(fraction * 100)}% of Item ${idx + 1} (Weight: ${currCap}/${w}, Value: ${Math.floor(v * fraction)}). Capacity full.`
      })
      break
    }
  }

  steps.push({
    array: [60, 100, 120],
    comparing: [],
    swapped: false,
    sorted: [0, 1, 2],
    line: 8,
    description: `Fractional Knapsack Complete! Total Maximum Value: ${totalVal}`
  })

  return steps
}

export function generateBoatsToSavePeopleSteps(people = [3, 2, 2, 1], limit = 3) {
  const peopleSorted = [...people].sort((a, b) => a - b)
  const n = peopleSorted.length
  let left = 0
  let right = n - 1
  let boats = 0
  const steps = []

  steps.push({
    array: [...peopleSorted],
    comparing: [left, right],
    swapped: false,
    line: 1,
    description: `Boats to Save People (LeetCode 881): Sorted people weights ${peopleSorted.join(', ')}, Boat Limit = ${limit}`
  })

  while (left <= right) {
    if (left === right) {
      boats++
      steps.push({
        array: [...peopleSorted],
        comparing: [left],
        swapped: true,
        line: 3,
        description: `Boat #${boats}: Single remaining person weight ${peopleSorted[left]} placed in boat.`
      })
      break
    }

    if (peopleSorted[left] + peopleSorted[right] <= limit) {
      boats++
      steps.push({
        array: [...peopleSorted],
        comparing: [left, right],
        swapped: true,
        line: 5,
        description: `Boat #${boats}: Paired Person ${peopleSorted[left]} + Person ${peopleSorted[right]} = ${peopleSorted[left] + peopleSorted[right]} <= ${limit}`
      })
      left++
      right--
    } else {
      boats++
      steps.push({
        array: [...peopleSorted],
        comparing: [right],
        swapped: false,
        line: 7,
        description: `Boat #${boats}: Heaviest person weight ${peopleSorted[right]} exceeds pair limit. Placed alone.`
      })
      right--
    }
  }

  steps.push({
    array: [...peopleSorted],
    comparing: [],
    swapped: false,
    sorted: Array.from({ length: n }, (_, i) => i),
    line: 10,
    description: `Greedy Boats Assignment Complete! Minimum Boats Required: ${boats}`
  })

  return steps
}

export function generateStonePileSteps(piles = [5, 4, 9]) {
  const pilesCurr = [...piles]
  const steps = []

  steps.push({
    array: [...pilesCurr],
    comparing: [],
    swapped: false,
    line: 1,
    description: `Stone Piles (LeetCode 1962): Initial stone piles ${pilesCurr.join(', ')}`
  })

  for (let stepNum = 1; stepNum <= 3; stepNum++) {
    const maxVal = Math.max(...pilesCurr)
    const maxIdx = pilesCurr.indexOf(maxVal)
    const reducedVal = maxVal - Math.floor(maxVal / 2)
    pilesCurr[maxIdx] = reducedVal

    steps.push({
      array: [...pilesCurr],
      comparing: [maxIdx],
      swapped: true,
      line: 4,
      description: `Greedy Operation #${stepNum}: Picked max pile ${maxVal} at index ${maxIdx}. Reduced to ${reducedVal} stones.`
    })
  }

  steps.push({
    array: [...pilesCurr],
    comparing: [],
    swapped: false,
    sorted: Array.from({ length: pilesCurr.length }, (_, i) => i),
    line: 8,
    description: `Stone Piles Minimization Complete! Final total stones: ${pilesCurr.reduce((a, b) => a + b, 0)}`
  })

  return steps
}

File: backend/algorithms/greedy.py
# Python Greedy Algorithms Step Generator

def generate_activity_selection_steps(start_times=None, finish_times=None):
    if start_times is None:
        start_times = [1, 3, 0, 5, 8, 5]
    if finish_times is None:
        finish_times = [2, 4, 6, 7, 9, 9]

    n = len(start_times)
    activities = sorted(range(n), key=lambda i: finish_times[i])
    selected = []

    steps = [{
        'array': finish_times,
        'comparing': [],
        'swapped': False,
        'description': 'Activity Selection: Sorting activities by finish time.'
    }]

    last_finish = -1
    for idx in activities:
        st = start_times[idx]
        ft = finish_times[idx]
        if st >= last_finish:
            selected.append(idx)
            last_finish = ft
            steps.append({
                'array': list(finish_times),
                'comparing': [idx],
                'swapped': True,
                'sorted': list(selected),
                'description': f'Selected Activity {idx + 1} (Start: {st}, Finish: {ft})'
            })
        else:
            steps.append({
                'array': list(finish_times),
                'comparing': [idx],
                'swapped': False,
                'sorted': list(selected),
                'description': f'Skipped Activity {idx + 1} (Start: {st} < Last Finish: {last_finish})'
            })

    steps.append({
        'array': list(finish_times),
        'comparing': [],
        'swapped': False,
        'sorted': list(selected),
        'description': f'Greedy Activity Selection Complete! Selected {len(selected)} non-overlapping activities.'
    })
    return steps


def generate_fractional_knapsack_steps(capacity=50):
    weights = [10, 20, 30]
    values = [60, 100, 120]
    ratios = [v / w for v, w in zip(values, weights)]

    steps = [{
        'array': [int(r) for r in ratios],
        'comparing': [],
        'swapped': False,
        'description': 'Fractional Knapsack: Calculate Value-to-Weight ratios.'
    }]

    total_val = 0.0
    curr_cap = capacity

    items = sorted(range(len(weights)), key=lambda i: ratios[i], reverse=True)
    for idx in items:
        w, v = weights[idx], values[idx]
        if w <= curr_cap:
            curr_cap -= w
            total_val += v
            steps.append({
                'array': [60, 100, 120],
                'comparing': [idx],
                'swapped': True,
                'description': f'Took 100% of Item {idx+1} (Weight: {w}, Value: {v}). Remaining Capacity: {curr_cap}'
            })
        else:
            fraction = curr_cap / w
            total_val += v * fraction
            steps.append({
                'array': [60, 100, 120],
                'comparing': [idx],
                'swapped': True,
                'description': f'Took {int(fraction*100)}% of Item {idx+1} (Weight: {curr_cap}/{w}, Value: {v * fraction:.1f}). Capacity full.'
            })
            break

    steps.append({
        'array': [60, 100, 120],
        'comparing': [],
        'swapped': False,
        'sorted': [0, 1, 2],
        'description': f'Fractional Knapsack Complete! Total Maximum Value: {total_val:.1f}'
    })
    return steps


def generate_boats_to_save_people_steps(people=None, limit=3):
    if people is None:
        people = [3, 2, 2, 1]

    people_sorted = sorted(people)
    n = len(people_sorted)
    left, right = 0, n - 1
    boats = 0

    steps = [{
        'array': list(people_sorted),
        'comparing': [left, right],
        'swapped': False,
        'description': f'Boats to Save People (LeetCode 881): Sorted people weights {people_sorted}, Boat Limit = {limit}'
    }]

    while left <= right:
        if left == right:
            boats += 1
            steps.append({
                'array': list(people_sorted),
                'comparing': [left],
                'swapped': True,
                'description': f'Boat #{boats}: Single remaining person weight {people_sorted[left]} placed in boat.'
            })
            break

        if people_sorted[left] + people_sorted[right] <= limit:
            boats += 1
            steps.append({
                'array': list(people_sorted),
                'comparing': [left, right],
                'swapped': True,
                'description': f'Boat #{boats}: Paired Person {people_sorted[left]} + Person {people_sorted[right]} = {people_sorted[left] + people_sorted[right]} <= {limit}'
            })
            left += 1
            right -= 1
        else:
            boats += 1
            steps.append({
                'array': list(people_sorted),
                'comparing': [right],
                'swapped': False,
                'description': f'Boat #{boats}: Heaviest person weight {people_sorted[right]} exceeds pair limit. Placed alone.'
            })
            right -= 1

    steps.append({
        'array': list(people_sorted),
        'comparing': [],
        'swapped': False,
        'sorted': list(range(n)),
        'description': f'Greedy Boats Assignment Complete! Minimum Boats Required: {boats}'
    })
    return steps


def generate_stone_pile_steps(piles=None):
    if piles is None:
        piles = [5, 4, 9]

    piles_curr = list(piles)
    steps = [{
        'array': list(piles_curr),
        'comparing': [],
        'swapped': False,
        'description': f'Stone Piles (LeetCode 1962): Initial stone piles {piles_curr}'
    }]

    for step_num in range(1, 4):
        max_idx = piles_curr.index(max(piles_curr))
        orig_val = piles_curr[max_idx]
        reduced_val = orig_val - (orig_val // 2)
        piles_curr[max_idx] = reduced_val

        steps.append({
            'array': list(piles_curr),
            'comparing': [max_idx],
            'swapped': True,
            'description': f'Greedy Operation #{step_num}: Picked max pile {orig_val} at index {max_idx}. Reduced to {reduced_val} stones.'
        })

    steps.append({
        'array': list(piles_curr),
        'comparing': [],
        'swapped': False,
        'sorted': list(range(len(piles_curr))),
        'description': f'Stone Piles Minimization Complete! Final total stones: {sum(piles_curr)}'
    })
    return steps
